File: learning-akka/examples/src/test/java/book/java/ch2/ReverseStringActorTest.java
package book.java.ch2;

import akka.actor.ActorRef;
import akka.actor.ActorSystem;
import akka.actor.Props;
import book.kotlin.util.ActorExtensions;
import org.junit.Test;

import java.util.Arrays;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutionException;
import java.util.stream.Stream;

import static book.kotlin.util.ActorExtensions.*;
import static java.util.stream.Collectors.toList;
import static org.junit.Assert.assertEquals;

public class ReverseStringActorTest {

	private ActorSystem system = ActorSystem.create();

	private ActorRef reverseActor = system.actorOf(Props.create(ReverseStringActor.class));

	@Test
	public void shouldReverseString() {
		CompletableFuture<String> future = ask(reverseActor, "123", 1000);

		assertEquals("321", block(future, 1000));
	}

	@Test(expected = ExecutionException.class)
	public void shouldFailWhenUnknownMessageSent() {
		CompletableFuture<Object> future = ask(reverseActor, 123, 1000);

		block(future, 1000);
	}

	@Test
	public void shouldRevertMultipleMessages() {
		List<CompletableFuture<String>> futures = Stream.of("123", "abc", "aaa")
				.map(s -> ActorExtensions.<String>ask(reverseActor, s, 1000))
				.collect(toList());
		CompletableFuture<List<String>> result = sequence(futures);
		System.out.println("RESULT " + result);
		assertEquals(Arrays.asList("321", "cba", "aaa"), block(result, 1000));
	}

}

File: learning-akka/examples/src/test/kotlin/book/kotlin/ch2/ReverseStringActorTest.kt
package book.kotlin.ch2

import akka.actor.ActorSystem
import book.kotlin.util.Props
import book.kotlin.util.ask
import book.kotlin.util.block
import book.kotlin.util.sequence
import org.junit.Assert.assertEquals
import org.junit.Test
import java.util.*
import java.util.concurrent.ExecutionException

class ReverseStringActorTest {

	private val system = ActorSystem.create()

	private val reverseActor = system.actorOf(Props<ReverseStringActor>())

	@Test
	fun `should reverse string`() {
		val future = reverseActor.ask<String>("123", 1000)

		assertEquals("321", future.block(1000))
	}

	@Test(expected = ExecutionException::class)
	fun `should fail when unknown message sent`() {
		val future = reverseActor.ask<Any>(123, 1000)

		future.block(1000)
	}

	@Test
	fun `should reverse multiple messages`() {
		val futures = listOf("123", "abc", "aaa").map { reverseActor.ask<String>(it, 1000) }
		val result = futures.sequence<String>()

		assertEquals(Arrays.asList("321", "cba", "aaa"), result.block(1000))
	}

}


